### PR TITLE
Added RegulatoryConfig support to AutoCommissioner for Android

### PR DESCRIPTION
**Pull request by swan-amazon**
> 5. Commissioner SHALL configure regulatory information in the Commissionee if it has at least one instance of Network Commissioning cluster on any endpoint with either the WI (i.e. Wi-Fi) or TH (i.e. Thread) feature flags set in its FeatureMap. Commissioner SHOULD configure UTC time, timezone, and DST offset, if the Commissionee supports the time cluster. The order of configuration of this information is not critical. The UTC time is configured using SetUtcTime command (see Section 11.16.7.1, “SetUtcTime Command”) while timezone and DST offset are set through TimeZone (see Section 11.16.6.6, “TimeZone Attribute”) and DstOffset attribute (see Section 11.16.6.7, “DstOffset Attribute”), respectively. The regulatory information is configured using SetRegulatoryConfig (see Section 11.9.7.4, “SetRegulatoryConfig Command”).

The CommissioningParameters has fields to support CountryCode and DeviceRegulatoryLocation, but the Android commissioner interface does not currently expose a way to set these fields. The interface has been updated to allow the commissioner to set these fields through the ControllerParams.

### CODE REVIEW

**Review by cliffamzn — APPROVED** [review]

File: src/controller/AutoCommissioner.cpp [hunk 1]
```diff
@@ -154,6 +154,18 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
...
+        ChipLogProgress(Controller, "Setting country code from parameters");
+        mParams.SetCountryCode(params.GetCountryCode().Value());
+    }
+
+    if (params.GetDeviceRegulatoryLocation().HasValue())
+    {
+        ChipLogProgress(Controller, "Setting regulatory location type from parameters");
+        mParams.SetDeviceRegulatoryLocation(params.GetDeviceRegulatoryLocation().Value());
+    }
+
```
Comment (bzbarsky-apple) [inline src/controller/AutoCommissioner.cpp:168]: Why is this needed, given the `mParams = params;` at the top of this function?

It seems there's a bunch of stuff in this method that should not be here; the only parts that really need to be are  the ones that change which buffers mParams points to.  All the things that get copied by that assignment should just be removed (in a separate PR).

Comment (bzbarsky-apple) [reply src/controller/AutoCommissioner.cpp:168]: I created https://github.com/project-chip/connectedhomeip/pull/24422 to remove the unnecessary parts here.

Comment (swan-amazon) [reply src/controller/AutoCommissioner.cpp]: I removed the get then set here. I think your PR should work, at least it seemed to work when I pulled and tested, but it was not an exhaustive test.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/controller/java/AndroidDeviceControllerWrapper.cpp [hunk 2]
```diff
@@ -92,7 +102,8 @@ AndroidDeviceControllerWrapper * AndroidDeviceControllerWrapper::AllocateNew(
 #endif
     jobject keypairDelegate, jbyteArray rootCertificate, jbyteArray intermediateCertificate, jbyteArray nodeOperationalCertificate,
     jbyteArray ipkEpochKey, uint16_t listenPort, uint16_t controllerVendorId, uint16_t failsafeTimerSeconds,
-    bool attemptNetworkScanWiFi, bool attemptNetworkScanThread, bool skipCommissioningComplete, CHIP_ERROR * errInfoOnFailure)
+    bool attemptNetworkScanWiFi, bool attemptNetworkScanThread, bool skipCommissioningComplete, jstring countryCode,
+    jint regulatoryLocationTypeValue, CHIP_ERROR * errInfoOnFailure)
```
Comment (tcarmelveilleux) [inline src/controller/java/AndroidDeviceControllerWrapper.cpp:106]: Do not add new arguments here. Use setters on the object to update the commissioning parameters. We cannot keep adding more args to this method

**Review by tcarmelveilleux — COMMENTED** [review]
- [ ] Please find a way to avoid adding yet one more argument to the already-too-long list of AllocateNew arguments (which also breaks every existing client of the SDK APIs on Android).  See https://github.com/project-chip/connectedhomeip/pull/24381 for a PR that adds new features to AndroidDeviceControllerWrapper without more arguments

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
- [ ] Please find a way to avoid adding yet one more argument to the already-too-long list of AllocateNew arguments (which also breaks every existing client of the SDK APIs on Android). See https://github.com/project-chip/connectedhomeip/pull/24381 for a PR that adds new features to AndroidDeviceControllerWrapper without more arguments

**Review by yunhanw-google — CHANGES_REQUESTED** [review]
please also add the android test for RegulatoryConfig.

https://github.com/project-chip/connectedhomeip/tree/master/examples/java-matter-controller

some instructions are as below.
-- Build
docker run -it -v /workspace/connectedhomeip:/connectedhomeip connectedhomeip/chip-build:0.6.06
git config --global --add [safe.directory](http://safe.directory/) /connectedhomeip
git config --global --add [safe.directory](http://safe.directory/) /connectedhomeip/third_party/pigweed/repo
export JAVA_PATH=/usr/lib/jvm/java-8-openjdk-amd64/
source scripts/bootstrap.sh
./scripts/build/build_examples.py --target linux-x64-java-matter-controller build
./scripts/build/build_examples.py --target linux-x64-all-clusters build

--In Terminal A
docker exec -it c20375aeb713 bash
rm -rf /tmp/chip* & ./out/linux-x64-all-clusters/chip-all-clusters-app

--In terminal B
docker exec -it c20375aeb713 bash
java -verbose:jni -Xcheck:jni -jar -Djava.library.path=out/linux-x64-java-matter-controller/lib/jni out/linux-x64-java-matter-controller/bin/java-matter-controller pairing onnetwork-long 1 20202021 3840 1000

File: src/controller/java/AndroidDeviceControllerWrapper.cpp (hunk 2 above)
Comment (swan-amazon) [reply src/controller/java/AndroidDeviceControllerWrapper.cpp]: I updated the PR to use setters and avoid additional parameters to the allocX function

**Comment by swan-amazon** [comment]
> please also add the android test for RegulatoryConfig.
> 
> https://github.com/project-chip/connectedhomeip/tree/master/examples/java-matter-controller
> 
> some instructions are as below. -- Build docker run -it -v /workspace/connectedhomeip:/connectedhomeip connectedhomeip/chip-build:0.6.06 git config --global --add [safe.directory](http://safe.directory/) /connectedhomeip git config --global --add [safe.directory](http://safe.directory/) /connectedhomeip/third_party/pigweed/repo export JAVA_PATH=/usr/lib/jvm/java-8-openjdk-amd64/ source scripts/bootstrap.sh ./scripts/build/build_examples.py --target linux-x64-java-matter-controller build ./scripts/build/build_examples.py --target linux-x64-all-clusters build
> 
> --In Terminal A docker exec -it c20375aeb713 bash rm -rf /tmp/chip* & ./out/linux-x64-all-clusters/chip-all-clusters-app
> 
> --In terminal B docker exec -it c20375aeb713 bash java -verbose:jni -Xcheck:jni -jar -Djava.library.path=out/linux-x64-java-matter-controller/lib/jni out/linux-x64-java-matter-controller/bin/java-matter-controller pairing onnetwork-long 1 20202021 3840 1000

After addressing Tennessee's comment, I ended up only making changes to the -JNI file, except around 10 trivial lines added to the ControllerParams builder.

Is this still a valid request?

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 3]
```diff
@@ -385,6 +397,52 @@ JNI_METHOD(jlong, newDeviceController)(JNIEnv * env, jobject self, jobject contr
...
+
+        if (countryCode != nullptr)
+        {
+            jstring countryCodeStr = static_cast<jstring>(countryCode);
+            JniUtfString countryCodeJniString(env, countryCodeStr);
+
+            VerifyOrExit(countryCodeJniString.size() == 2, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+            chip::Controller::CommissioningParameters commissioningParams = wrapper->GetCommissioningParameters();
+            commissioningParams.SetCountryCode(countryCodeJniString.charSpan());
```
Comment (tcarmelveilleux) [inline src/controller/java/CHIPDeviceController-JNI.cpp:413]: >     // The lifetime of the buffer countryCode is pointing to should exceed the
>     // lifetime of CommissioningParameters object.

^^^ You must add a member in the wrapper that is a char[2], and copy from the Java object to that storage, and then call SetCountryCode. Otherwise, the countryCodeJniString string will go out of scope and the span will point to garbage.

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 4]
```diff
@@ -385,6 +397,52 @@ JNI_METHOD(jlong, newDeviceController)(JNIEnv * env, jobject self, jobject contr
...
+            err = wrapper->UpdateCommissioningParameters(commissioningParams);
```
Comment (tcarmelveilleux) [inline src/controller/java/CHIPDeviceController-JNI.cpp:414]: Please add a comment that indicates that `UpdateCommissioningParameters` has owning storage for the country code, which is why it's OK to just set a Span on SetCountryCode backed by an object that is transient

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 5]
```diff
@@ -385,6 +397,52 @@ JNI_METHOD(jlong, newDeviceController)(JNIEnv * env, jobject self, jobject contr
...
+            jint regulatoryLocationJint = chip::JniReferences::GetInstance().IntegerToPrimitive(regulatoryLocation);
+            VerifyOrExit(chip::CanCastTo<RegulatoryLocationType>(regulatoryLocationJint), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+            RegulatoryLocationType regulatoryLocationType = static_cast<RegulatoryLocationType>(regulatoryLocationJint);
+            VerifyOrExit(regulatoryLocationType >= RegulatoryLocationType::kIndoor, err = CHIP_ERROR_INVALID_ARGUMENT);
+            VerifyOrExit(regulatoryLocationType <= RegulatoryLocationType::kIndoorOutdoor, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+            chip::Controller::CommissioningParameters commissioningParams = wrapper->GetCommissioningParameters();
+            commissioningParams.SetDeviceRegulatoryLocation(regulatoryLocationType);
+            err = wrapper->UpdateCommissioningParameters(commissioningParams);
```
Comment (tcarmelveilleux) [inline src/controller/java/CHIPDeviceController-JNI.cpp:439]: UpdateCommissioningParameters has different issues here....
 
>     // this will wipe out any custom attestationNonce and csrNonce that was being used.
>     // however, Android APIs don't allow these to be set to custom values today.

You are following the pattern already existing, but it strikes me that this entire method has been built via cut and paste of prior "similar" blocks in a way that by now, I feel like a variety of complex ownership issues exist for the commissioning parameters... Won't block your PR, but the realization I made here is quite concerning...

Comment (bzbarsky-apple) [reply src/controller/java/CHIPDeviceController-JNI.cpp:439]: Fwiw, I don't even think that comment is true. Though there _is_ a possible issue in doing `memcpy` from a pointer to itself, not a `memmove`, but I think if the pointers are in fact equal it might mostly turn out ok...

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 6]
```diff
@@ -384,6 +396,54 @@ JNI_METHOD(jlong, newDeviceController)(JNIEnv * env, jobject self, jobject contr
...
+
+        if (countryCode != nullptr)
+        {
+            jstring countryCodeStr = static_cast<jstring>(countryCode);
+            JniUtfString countryCodeJniString(env, countryCodeStr);
+
+            VerifyOrExit(countryCodeJniString.size() == 2, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+            chip::Controller::CommissioningParameters commissioningParams = wrapper->GetCommissioningParameters();
+            commissioningParams.SetCountryCode(countryCodeJniString.charSpan());
```
Comment (tcarmelveilleux) [inline src/controller/java/CHIPDeviceController-JNI.cpp:412]: SetCountryCode sets a span. You need to have an owned copy of country code and regulatory location in the wrapper (not a span, but the actual values).

**Review by tcarmelveilleux — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]
